ianalyzer_readers/readers/core.py

```diff
@@ -142,20 +174,172 @@ def source2dicts(self, source: Source) -> Iterable[Document]:
+            data = self.data_from_file(source_data)
+        elif isinstance(source, bytes):
+            data = self.data_from_bytes(source_data)
+        elif isinstance(source, Response):
+            data = self.data_from_response(source_data)
+        else:
+            raise TypeError(f'Unknown source type: {type(source_data)}')
+
+        return data, metadata
+
+
+    def data_from_file(self, path: str) -> Any:
+        '''
+        Extract source data from a filename.
+
+        The return type depends on how the reader is implemented, but should be some kind
+        of data structure from which documents can be extracted. It serves as the input
+        to `self.iterate_data`.
+
+        This method can also return a context manager. This is especially useful to
+        iterate over large files in `iterate_data`, without loading the complete file
+        contents in memory.
 
+        Tip: if you have implemented `self.data_from_bytes`, this method can probably just
+        read the binary contents of the file and call that method.
+
+        Parameters:
+            path: The path to a file.
+        
+        Returns:
+            A data object. The type depends on the reader implementation.
+        
         Raises:
-            NotImplementedError: This method needs to be implemented on child
-                classes. It will raise an error by default.
+            NotImplementedError: this method may be implemented on child classes, but
+                has no default implementation.
+        '''
+        
+        raise NotImplementedError('This reader does not support filename input')
+
+
+    def data_from_bytes(self, bytes: bytes) -> Any:
+        '''
+        Extract source data from a bytes object. Like `data_from_file`, but with bytes
+        input.
+
+        Parameters:
+            bytes: byte contents of the source
+        
+        Returns:
+            A data object. The type depends on the reader implementation. This may also
+                be a context manager.
+        
+        Raises:
+            NotImplementedError: this method may be implemented on child classes, but
+                has no default implementation.
+        '''
+        
+        raise NotImplementedError('This reader does not support bytes input')
+
+
+    def data_from_response(self, response: Response) -> Any:
+        '''
+        Extract data from an HTTP response. Like `data_from_file`, but with `Response`
+        input.
+
+        Parameters:
+            response: HTTP response object
+        
+        Returns:
+            A data object. The type depends on the reader implementation. This may also
+                be a context manager.
+        
+        Raises:
+            NotImplementedError: this method may be implemented on child classes, but has
+                no default implementation.
+        '''
+        raise NotImplementedError('This reader does not support Response input')
```

Comment:
Would this not cause hard failures when attempting to index with most reader classes? Few of them implement reading from bytes, and only one sof far implements reading from `Response` objects.

Reply:
It would, if you would create, say, a subclass of `CSVReader` (which doesn't implement this, I think) and then return `Response` objects in `sources()`. This was already true: you can only process sources that your reader supports.

New to this version is that you could fix the problem by implementing `data_from_response` in your child class. In the current version, you could only fix the problem by rewriting `source2dicts` in its entirety.